partiql-types/src/main/kotlin/org/partiql/value/datetime/DateTime.kt

```diff
@@ -0,0 +1,560 @@
+
+    public override fun hashCode(): Int
+
+    public override fun toString(): String
+}
+
+/**
+ * Superclass for all implementations representing date value.
+ * Date represents a calendar system, (i.e., 2023-06-01).
+ * It does not include information on time or timezone, instead, it is meant to represent a specific date on calendar.
+ * For example, 2022-11-25 (black friday in 2022).
+ * The valid range are from 0001-01-01 to 9999-12-31
+ * The [day] must be valid for the year and month, otherwise an exception will be thrown.
+ */
+public sealed interface Date : DateTime, Comparable<Date> {
+
+    public override val year: Int
+    public override val month: Int
+    public override val day: Int
+
+    /**
+     * Hour field for [Date] value is always null.
+     */
+    public override val hour: Int?
+        get() = null
+
+    /**
+     * Minute field for [Date] value is always null.
+     */
+    public override val minute: Int?
+        get() = null
+
+    /**
+     * Second field for [Date] value is always null.
+     */
+    public override val decimalSecond: BigDecimal?
+        get() = null
+
+    /**
+     * Timezone field for [Date] value is always null.
+     */
+    public override val timeZone: TimeZone?
+        get() = null
+
+    // Operation
+    /**
+     * Construct a [Timestamp] value by appending [time] to this [Date] value.
+     */
+    public fun atTime(time: Time): Timestamp
+
+    /**
+     * Returns a [Date] value with the specified number of days added.
+     * The month and year fields may be changed as necessary to ensure the result remains valid.
+     * [days] can be negative.
+     */
+    public fun plusDays(days: Long): Date
+
+    /**
+     * Returns a [Date] value with the specified number of months added.
+     * The month and year fields may be changed as necessary to ensure the result remains valid.
+     * [months] can be negative.
+     */
+    public fun plusMonths(months: Long): Date
+
+    /**
+     * Returns a [Date] value with the specified number of months added.
+     * [years] can be negative.
+     */
+    public fun plusYear(years: Long): Date
+
+    /**
+     * Comparison method for [Date] value.
+     *
+     * Since [Date] value has no concept of time zone, they are compared as calendar date.
+     */
+    public override fun compareTo(other: Date): Int =
+        DateTimeComparator.compareTo(this, other)
+}
+
+/**
+ * Superclass for all implementations representing time value.
+ * See [TimeWithTimeZone] and [TimeWithoutTimeZone]
+ */
+public sealed interface Time : DateTime, Comparable<Time> {
+    /**
+     * Year field for [Time] is always null
+     */
+    public override val year: Int?
+        get() = null
+
+    /**
+     * Month field for [Time] is always null
+     */
+    public override val month: Int?
+        get() = null
+
+    /**
+     * Day field for [Time] is always null
+     */
+    public override val day: Int?
+        get() = null
```

Comment:
Why getters? Can you just assign null?

Reply:
Property initializers are not allowed in interfaces.